runtime/etc/Makefile.mli-shared

```diff
@@ -0,0 +1,140 @@
+# Copyright 2004-2019 Cray Inc.
+# Other additional copyright holders may be indicated within.
+#
+# The entirety of this work is licensed under the Apache License,
+# Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License.
+#
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+#
+# Makefile.mli-static
+#
+
+include $(CHPL_MAKE_HOME)/runtime/etc/Makefile.include
+
+CHPL_MLI_INCS = -I$(CHPL_ZMQ_LOC)/include
+CHPL_LN_LIB_DIR = $(CHPL_MAKE_RUNTIME_LIB)/$(CHPL_MAKE_LAUNCHER_SUBDIR)
+UNAME = $(shell uname)
+
+.PHONY: checkZMQ buildServer buildClient runTagsCommand
+
+all: checkZMQ buildServer buildClient runTagsCommand FORCE
+
+checkZMQ:
+ifndef CHPL_ZMQ_LOC
+  $(error CHPL_ZMQ_LOC must be set to the location of the ZeroMQ library)
+endif
+
+buildServer: $(TMPSERVERNAME)
+
+buildClient: $(TMPBINNAME)
+
+runTagsCommand:
+	$(TAGS_COMMAND)
+
+$(TMPBINNAME): $(CHPL_CL_OBJS) FORCE
+#
+# Invoke the launcher script, but don't actually link up the launcher binary.
+# Do this by setting CHPL_MAKE_IS_MLI=1.
+#
+	$(MAKE) -f $(CHPL_MAKE_HOME)/runtime/etc/Makefile.launcher all \
+		CHPL_MAKE_HOME=$(CHPL_MAKE_HOME) \
+		TMPBINNAME=$(TMPSERVERNAME) \
+		BINNAME=$(SERVERNAME) \
+		TMPDIRNAME=$(TMPDIRNAME) \
+		CHPL_MAKE_RUNTIME_LIB=$(CHPL_MAKE_RUNTIME_LIB) \
+		CHPL_MAKE_RUNTIME_INCL=$(CHPL_MAKE_RUNTIME_INCL) \
+		CHPL_MAKE_THIRD_PARTY=$(CHPL_MAKE_THIRD_PARTY) \
+		CHPL_MAKE_IS_MLI=1
+#
+# Compile the client bundle. Exactly the same as for static compilation, the
+# only thing that changes here are the flags inserted by codegen_makefile.
```

Comment:
One of the things that makes it hard to understand exactly how a Chapel program gets built is the splitting of responsibility between two places which are very far apart: the compiler's back end, and the Makefiles stored in an obscure corner of the `$CHPL_HOME/runtime` hierarchy.  We shouldn't presume that a reader of one is necessarily familiar with the other, so perhaps instead of just saying "codegen_makefile" here, say "codegen_makefile() in the compiler"?

It would be nice if all of the program build logic could be together in one place.  That's not 100% practical, but it's something to strive for.  In the case of `Makefile.mli-static` and `Makefile.mli-dynamic` we know whether we're linking statically or dynamically, so it's not actually necessary for the compiler to put the linking mode option into the args to be passed into the Makefile.  `Makefile.mli-static` and `Makefile.mli-dynamic` could themselves throw `$LIB_STATIC_FLAG` and `$LIB_DYNAMIC_FLAG`, respectively.  Basically we would be moving the logic currently at files.cpp:877 (in this PR) into `Makefile.mli-*`. The resulting small decrease in compiler/Makefile interdependence seems like a plus to me.

Reply:
That is a really good idea that I've considered myself. Since I "own" these mli-* Makefiles, I suppose I can do this.